compiler/resolution/resolveFunction.cpp

```diff
@@ -574,6 +588,233 @@ static void markTypesWithDefaultInitEqOrAssign(FnSymbol* fn) {
+      USR_FATAL_CONT(have.assign->defPoint,
+                     "Can't overload assignments for class types");
+    }
+  }
+
+  // Don't worry about checking for 'init=' for tuples or types
+  // with runtime type since these do not use 'init=' in the usual way.
+  if (toType->symbol->hasFlag(FLAG_TUPLE) ||
+      toType->symbol->hasFlag(FLAG_HAS_RUNTIME_TYPE)) {
+    checkInitEq = false;
+    // However, don't allow '=' on these types to be defined
+    // outside of the standard/internal modules
+    if (have.assign != NULL &&
+        have.assign->defPoint->getModule()->modTag == MOD_USER) {
+      USR_FATAL_CONT(have.assign->defPoint,
+                     "unsupported = overload to '%s' from '%s'",
+                     toString(toType), toString(fromType));
+      USR_PRINT("assignment overloads setting array or tuples "
+                "are not currently supported");
+      checkAssign = false;
+      checkCast = false;
+    }
+  }
+
+  // Don't look for casts to array/domain types
+  // (these should probably be added, right?)
+  if (toType->symbol->hasFlag(FLAG_HAS_RUNTIME_TYPE)) {
+    checkCast = false;
+  }
+
+  // Don't look for casts or init= to iterators. These don't make sense.
+  if (toType->symbol->hasFlag(FLAG_ITERATOR_RECORD)) {
+    checkCast = false;
+    checkInitEq = false;
+  }
+
+  // Since we don't yet have init= on non-records or tertiary records,
+  // don't worry about init= yet for non-records or extern types.
+  if (!(isRecord(toType) || isUnion(toType)) ||
+      toType->symbol->hasFlag(FLAG_EXTERN))
+    checkInitEq = false;
+
+  // add calls to resolve these if not already known
+  if (checkAssign && have.assign == NULL) {
+    // also resolve =
+    BlockStmt* block = new BlockStmt(BLOCK_SCOPELESS);
+    forCall->insertBefore(block);
+
+    VarSymbol* toTmp = newTemp("_check_to", toType);
+    VarSymbol* fromTmp = newTemp("_check_from", fromType);
+    CallExpr* c = new CallExpr(astrSassign, toTmp, fromTmp);
+    block->insertAtTail(new DefExpr(toTmp));
+    block->insertAtTail(new DefExpr(fromTmp));
+    block->insertAtTail(c);
+    tryResolveCall(c);
+    block->remove();
+    // update our local variable
+    // (above resolve call should have updated the map)
+    have.assign = c->resolvedFunction();
+  }
+
+  if (checkInitEq && have.initEq == NULL) {
+    // also resolve init=
+    BlockStmt* block = new BlockStmt(BLOCK_SCOPELESS);
+    forCall->insertBefore(block);
+
+    VarSymbol* toTmp = newTemp("_check_to", toType);
+    VarSymbol* fromTmp = newTemp("_check_from", fromType);
+    CallExpr* c = new CallExpr(astrInitEquals, gMethodToken, toTmp, fromTmp);
+    block->insertAtTail(new DefExpr(toTmp));
+    block->insertAtTail(new DefExpr(fromTmp));
+    block->insertAtTail(c);
+    tryResolveCall(c);
+    block->remove();
+    // update our local variable
+    // (above resolve call should have updated the map)
+    have.initEq = c->resolvedFunction();
```

Comment:
Could/should some of this code be shared with the previous branch?

Reply:
It's not clear to me how to do that without making the code much less straightforward (e.g. adding a function that adds temps and calls) so I'd like to keep it this way.